torch/csrc/jit/tensorexpr/mem_arena.h

```diff
@@ -23,20 +24,20 @@ class KernelArena {
 };
 
 // A RAII convenience wrapper on top of a kernel.
-// It either creates a Kernel, or take another existing Kernel, and sets it as
-// the current Kernel, as long as this KernelScope object is alive.
+// It creates a Kernel and sets it as the current Kernel, as long as this
+// KernelScope object is alive. After that, the previous Kernel is set as
+// current.
 class KernelScope {
  public:
   TORCH_API KernelScope();
-  TORCH_API explicit KernelScope(KernelArena& kernel_arena);
-  TORCH_API ~KernelScope() noexcept(false);
+  TORCH_API ~KernelScope();
 
  private:
   KernelScope(const KernelScope&) = delete;
   KernelScope& operator=(const KernelScope&) = delete;
-  bool owning_kernel_arena_ = false;
-  KernelArena* kernel_arena_ =
-      nullptr; // possibly owned, if owning_kernel_arena_ == true
+  KernelArena* kernel_arena_ = nullptr;     // owned, created in constructor
```

Comment:
Decouple KernelArena and KernelScope ownership is a useful feature. It is certainly possible to have an arena, and have it active and deactivated multiple times. On the other side, KernelScope is a convenience RAII wrapper that is supposed to be used only once to activate/deactivate the arena.

Reply:
> Decouple KernelArena and KernelScope ownership is a useful feature. It is certainly possible to have an arena, and have it active and deactivated multiple times. On the other side, KernelScope is a convenience RAII wrapper that is supposed to be used only once to activate/deactivate the arena.

In the long run I hope we would be able to completely get rid of scopes - using a global for the current arena is a workaround, what I think we should be doing is determining the arena when creating an expression from its operands or requiring it to be specified explicitly if it cannot be deduced from the operands.